File: baselibrary/src/main/java/com/minicart/android/baselibrary/dagger2/ActivityLifecycle.java
package com.minicart.android.baselibrary.dagger2;

import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentManager;
import android.view.View;

import com.minicart.android.baselibrary.dagger2.delegate.ActivityDelegate;
import com.minicart.android.baselibrary.dagger2.delegate.ActivityDelegateImpl;
import com.minicart.android.baselibrary.dagger2.delegate.FragmentDelegate;
import com.minicart.android.baselibrary.dagger2.delegate.FragmentDelegateImpl;
import com.minicart.android.baselibrary.dagger2.delegate.IActivity;
import com.minicart.android.baselibrary.dagger2.delegate.IFragment;
import com.minicart.android.baselibrary.support.AppManager;
import com.minicart.android.baselibrary.support.LogUtil;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class ActivityLifecycle implements Application.ActivityLifecycleCallbacks {

    private AppManager mAppManager;
    private Application mApplication;
    private FragmentLifecycle mFragmentLifecycle;

    @Inject
    public ActivityLifecycle(AppManager appManager, Application application) {
        this.mAppManager = appManager;
        this.mApplication = application;
    }

    @Override
    public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
        LogUtil.d(activity.getClass().getSimpleName(), "Created");
        mAppManager.pullActivity(activity);
        if (activity instanceof IActivity && activity.getIntent() != null) {
            ActivityDelegate activityDelegate = fetchActivityDelegate(activity);
            if (activityDelegate == null) {
                activityDelegate = new ActivityDelegateImpl(activity);
                activity.getIntent().putExtra(ActivityDelegate.ACTIVITY_DELEGATE, activityDelegate);
            }
            activityDelegate.onCreate(savedInstanceState);

            boolean useFragment = ((IActivity) activity).useFragment();
            if (activity instanceof FragmentActivity && useFragment) {
                if (mFragmentLifecycle == null) {
                    mFragmentLifecycle = new FragmentLifecycle();
                }
                ((FragmentActivity) activity).getSupportFragmentManager().registerFragmentLifecycleCallbacks(mFragmentLifecycle, true);
            }
        }
    }

    @Override
    public void onActivityStarted(Activity activity) {
        ActivityDelegate activityDelegate = fetchActivityDelegate(activity);
        if (activityDelegate != null) {
            activityDelegate.onStart();
        }
    }

    @Override
    public void onActivityResumed(Activity activity) {
        mAppManager.setCurrentActivity(activity);

        ActivityDelegate activityDelegate = fetchActivityDelegate(activity);
        if (activityDelegate != null) {
            activityDelegate.onResume();
        }
    }

    @Override
    public void onActivityPaused(Activity activity) {
        ActivityDelegate activityDelegate = fetchActivityDelegate(activity);
        if (activityDelegate != null) {
            activityDelegate.onPause();
        }
    }

    @Override
    public void onActivityStopped(Activity activity) {
        if (mAppManager.getCurrentActivity() == activity) {
            mAppManager.setCurrentActivity(null);
        }
        ActivityDelegate activityDelegate = fetchActivityDelegate(activity);
        if (activityDelegate != null) {
            activityDelegate.onStop();
        }
    }

    @Override
    public void onActivitySaveInstanceState(Activity activity, Bundle outState) {
        ActivityDelegate activityDelegate = fetchActivityDelegate(activity);
        if (activityDelegate != null) {
            activityDelegate.onSaveInstanceState(outState);
        }
    }

    @Override
    public void onActivityDestroyed(Activity activity) {
        mAppManager.removeActivity(activity);

        ActivityDelegate activityDelegate = fetchActivityDelegate(activity);
        if (activityDelegate != null) {
            activityDelegate.onDestroy();
            activity.getIntent().removeExtra(ActivityDelegate.ACTIVITY_DELEGATE);
        }
    }

    private ActivityDelegate fetchActivityDelegate(Activity activity) {
        ActivityDelegate activityDelegate = null;
        if (activity instanceof IActivity && activity.getIntent() != null) {
            activityDelegate = activity.getIntent().getParcelableExtra(ActivityDelegate.ACTIVITY_DELEGATE);
        }
        return activityDelegate;
    }

    static class FragmentLifecycle extends FragmentManager.FragmentLifecycleCallbacks {

        @Override
        public void onFragmentAttached(FragmentManager fm, Fragment f, Context context) {
            super.onFragmentAttached(fm, f, context);
            if (f instanceof IFragment && f.getArguments() != null) {
                FragmentDelegate fragmentDelegate = fetchFragmentDelegate(f);
                if (fragmentDelegate == null || !fragmentDelegate.isAdded()) {
                    fragmentDelegate = new FragmentDelegateImpl(fm, f);
                    f.getArguments().putParcelable(FragmentDelegate.FRAGMENT_DELEGATE, fragmentDelegate);
                }
                fragmentDelegate.onAttach(context);
            }
        }

        @Override
        public void onFragmentCreated(FragmentManager fm, Fragment f, Bundle savedInstanceState) {
            super.onFragmentCreated(fm, f, savedInstanceState);
            FragmentDelegate fragmentDelegate = fetchFragmentDelegate(f);
            if (fragmentDelegate != null) {
                fragmentDelegate.onCreate(savedInstanceState);
            }
        }

        @Override
        public void onFragmentViewCreated(FragmentManager fm, Fragment f, View v, Bundle savedInstanceState) {
            super.onFragmentViewCreated(fm, f, v, savedInstanceState);
            FragmentDelegate fragmentDelegate = fetchFragmentDelegate(f);
            if (fragmentDelegate != null) {
                fragmentDelegate.onCreateView(v, savedInstanceState);
            }
        }

        @Override
        public void onFragmentActivityCreated(FragmentManager fm, Fragment f, Bundle savedInstanceState) {
            super.onFragmentActivityCreated(fm, f, savedInstanceState);
            FragmentDelegate fragmentDelegate = fetchFragmentDelegate(f);
            if (fragmentDelegate != null) {
                fragmentDelegate.onActivityCreate(savedInstanceState);
            }
        }

        @Override
        public void onFragmentStarted(FragmentManager fm, Fragment f) {
            super.onFragmentStarted(fm, f);
            FragmentDelegate fragmentDelegate = fetchFragmentDelegate(f);
            if (fragmentDelegate != null) {
                fragmentDelegate.onStart();
            }
        }

        @Override
        public void onFragmentResumed(FragmentManager fm, Fragment f) {
            super.onFragmentResumed(fm, f);
            FragmentDelegate fragmentDelegate = fetchFragmentDelegate(f);
            if (fragmentDelegate != null) {
                fragmentDelegate.onResume();
            }
        }

        @Override
        public void onFragmentPaused(FragmentManager fm, Fragment f) {
            super.onFragmentPaused(fm, f);
            FragmentDelegate fragmentDelegate = fetchFragmentDelegate(f);
            if (fragmentDelegate != null) {
                fragmentDelegate.onPause();
            }
        }

        @Override
        public void onFragmentStopped(FragmentManager fm, Fragment f) {
            super.onFragmentStopped(fm, f);
            FragmentDelegate fragmentDelegate = fetchFragmentDelegate(f);
            if (fragmentDelegate != null) {
                fragmentDelegate.onStop();
            }
        }

        @Override
        public void onFragmentViewDestroyed(FragmentManager fm, Fragment f) {
            super.onFragmentViewDestroyed(fm, f);
            FragmentDelegate fragmentDelegate = fetchFragmentDelegate(f);
            if (fragmentDelegate != null) {
                fragmentDelegate.onDestroyView();
            }
        }

        @Override
        public void onFragmentDestroyed(FragmentManager fm, Fragment f) {
            super.onFragmentDestroyed(fm, f);
            FragmentDelegate fragmentDelegate = fetchFragmentDelegate(f);
            if (fragmentDelegate != null) {
                fragmentDelegate.onDestroy();
            }
        }

        @Override
        public void onFragmentDetached(FragmentManager fm, Fragment f) {
            super.onFragmentDetached(fm, f);
            FragmentDelegate fragmentDelegate = fetchFragmentDelegate(f);
            if (fragmentDelegate != null) {
                fragmentDelegate.onDetach();
                f.getArguments().clear();
            }
        }


        private FragmentDelegate fetchFragmentDelegate(Fragment fragment) {
            if (fragment instanceof IFragment) {
                return fragment.getArguments() == null ? null : (FragmentDelegate) fragment.getArguments().getParcelable(FragmentDelegate.FRAGMENT_DELEGATE);
            }
            return null;
        }
    }

}

File: baselibrary-kotlin/src/main/java/com/minicart/android/baselibrary/dagger2/ActivityLifecycle.kt
package com.minicart.android.baselibrary.dagger2

import android.app.Activity
import android.app.Application
import android.content.Context
import android.os.Bundle
import android.os.Parcelable
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentActivity
import android.support.v4.app.FragmentManager
import android.view.View

import com.minicart.android.baselibrary.dagger2.delegate.ActivityDelegate
import com.minicart.android.baselibrary.dagger2.delegate.ActivityDelegateImpl
import com.minicart.android.baselibrary.dagger2.delegate.FragmentDelegate
import com.minicart.android.baselibrary.dagger2.delegate.FragmentDelegateImpl
import com.minicart.android.baselibrary.dagger2.delegate.IActivity
import com.minicart.android.baselibrary.dagger2.delegate.IFragment
import com.minicart.android.baselibrary.support.AppManager
import com.minicart.android.baselibrary.support.LogUtil

import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ActivityLifecycle @Inject
constructor(private val mAppManager: AppManager, private val mApplication: Application) : Application.ActivityLifecycleCallbacks {
    private var mFragmentLifecycle: FragmentLifecycle? = null

    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle) {
        LogUtil.d(activity.javaClass.simpleName, "Created")
        mAppManager.pullActivity(activity)
        if (activity is IActivity<*> && activity.intent != null) {
            var activityDelegate = fetchActivityDelegate(activity)
            if (activityDelegate == null) {
                activityDelegate = ActivityDelegateImpl(activity)
                activity.intent.putExtra(ActivityDelegate.ACTIVITY_DELEGATE, activityDelegate)
            }
            activityDelegate.onCreate(savedInstanceState)

            val useFragment = (activity as IActivity<*>).useFragment()
            if (activity is FragmentActivity && useFragment) {
                if (mFragmentLifecycle == null) {
                    mFragmentLifecycle = FragmentLifecycle()
                }
                (activity as FragmentActivity).supportFragmentManager.registerFragmentLifecycleCallbacks(mFragmentLifecycle, true)
            }
        }
    }

    override fun onActivityStarted(activity: Activity) {
        val activityDelegate = fetchActivityDelegate(activity)
        activityDelegate?.onStart()
    }

    override fun onActivityResumed(activity: Activity) {
        mAppManager.currentActivity = activity

        val activityDelegate = fetchActivityDelegate(activity)
        activityDelegate?.onResume()
    }

    override fun onActivityPaused(activity: Activity) {
        val activityDelegate = fetchActivityDelegate(activity)
        activityDelegate?.onPause()
    }

    override fun onActivityStopped(activity: Activity) {
        if (mAppManager.currentActivity === activity) {
            mAppManager.currentActivity = null
        }
        val activityDelegate = fetchActivityDelegate(activity)
        activityDelegate?.onStop()
    }

    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {
        val activityDelegate = fetchActivityDelegate(activity)
        activityDelegate?.onSaveInstanceState(outState)
    }

    override fun onActivityDestroyed(activity: Activity) {
        mAppManager.removeActivity(activity)

        val activityDelegate = fetchActivityDelegate(activity)
        if (activityDelegate != null) {
            activityDelegate.onDestroy()
            activity.intent.removeExtra(ActivityDelegate.ACTIVITY_DELEGATE)
        }
    }

    private fun fetchActivityDelegate(activity: Activity): ActivityDelegate? {
        var activityDelegate: ActivityDelegate? = null
        if (activity is IActivity<*> && activity.intent != null) {
            activityDelegate = activity.intent.getParcelableExtra<ActivityDelegate>(ActivityDelegate.ACTIVITY_DELEGATE)
        }
        return activityDelegate
    }

    internal class FragmentLifecycle : FragmentManager.FragmentLifecycleCallbacks() {

        override fun onFragmentAttached(fm: FragmentManager?, f: Fragment?, context: Context?) {
            super.onFragmentAttached(fm, f, context)
            if (f is IFragment && f.arguments != null) {
                var fragmentDelegate = fetchFragmentDelegate(f)
                if (fragmentDelegate == null || !fragmentDelegate.isAdded) {
                    fragmentDelegate = FragmentDelegateImpl(fm!!, f)
                    f.arguments.putParcelable(FragmentDelegate.FRAGMENT_DELEGATE, fragmentDelegate)
                }
                fragmentDelegate.onAttach(context!!)
            }
        }

        override fun onFragmentCreated(fm: FragmentManager?, f: Fragment?, savedInstanceState: Bundle?) {
            super.onFragmentCreated(fm, f, savedInstanceState)
            val fragmentDelegate = fetchFragmentDelegate(f)
            fragmentDelegate?.onCreate(savedInstanceState!!)
        }

        override fun onFragmentViewCreated(fm: FragmentManager?, f: Fragment?, v: View?, savedInstanceState: Bundle?) {
            super.onFragmentViewCreated(fm, f, v, savedInstanceState)
            val fragmentDelegate = fetchFragmentDelegate(f)
            fragmentDelegate?.onCreateView(v!!, savedInstanceState!!)
        }

        override fun onFragmentActivityCreated(fm: FragmentManager?, f: Fragment?, savedInstanceState: Bundle?) {
            super.onFragmentActivityCreated(fm, f, savedInstanceState)
            val fragmentDelegate = fetchFragmentDelegate(f)
            fragmentDelegate?.onActivityCreate(savedInstanceState!!)
        }

        override fun onFragmentStarted(fm: FragmentManager?, f: Fragment?) {
            super.onFragmentStarted(fm, f)
            val fragmentDelegate = fetchFragmentDelegate(f)
            fragmentDelegate?.onStart()
        }

        override fun onFragmentResumed(fm: FragmentManager?, f: Fragment?) {
            super.onFragmentResumed(fm, f)
            val fragmentDelegate = fetchFragmentDelegate(f)
            fragmentDelegate?.onResume()
        }

        override fun onFragmentPaused(fm: FragmentManager?, f: Fragment?) {
            super.onFragmentPaused(fm, f)
            val fragmentDelegate = fetchFragmentDelegate(f)
            fragmentDelegate?.onPause()
        }

        override fun onFragmentStopped(fm: FragmentManager?, f: Fragment?) {
            super.onFragmentStopped(fm, f)
            val fragmentDelegate = fetchFragmentDelegate(f)
            fragmentDelegate?.onStop()
        }

        override fun onFragmentViewDestroyed(fm: FragmentManager?, f: Fragment?) {
            super.onFragmentViewDestroyed(fm, f)
            val fragmentDelegate = fetchFragmentDelegate(f)
            fragmentDelegate?.onDestroyView()
        }

        override fun onFragmentDestroyed(fm: FragmentManager?, f: Fragment?) {
            super.onFragmentDestroyed(fm, f)
            val fragmentDelegate = fetchFragmentDelegate(f)
            fragmentDelegate?.onDestroy()
        }

        override fun onFragmentDetached(fm: FragmentManager?, f: Fragment?) {
            super.onFragmentDetached(fm, f)
            val fragmentDelegate = fetchFragmentDelegate(f)
            if (fragmentDelegate != null) {
                fragmentDelegate.onDetach()
                f!!.arguments.clear()
            }
        }


        private fun fetchFragmentDelegate(fragment: Fragment): FragmentDelegate? {
            if (fragment is IFragment) {
                return if (fragment.arguments == null) null else fragment.arguments.getParcelable<Parcelable>(FragmentDelegate.FRAGMENT_DELEGATE)
            }
            return null
        }
    }

}